src/charm.py

```diff
@@ -1,233 +1,294 @@
+                    "override": "replace",
+                    "environment": context,
+                    "on-check-failure": {"up": "ignore"},
+                }
+            },
+            "checks": {
+                "up": {
+                    "override": "replace",
+                    "period": "10s",
+                    "http": {"url": f"http://localhost:{OAUTH2_PROXY_API_PORT}/ready"},
+                }
+            },
+        }
+        return Layer(layer_config)
+
+    def _validate_pebble_plan(self) -> bool:
+        """Validate pebble plan."""
+        try:
+            plan = self._container.get_plan().to_dict()
+            return bool(plan["services"][WORKLOAD_SERVICE]["on-check-failure"])
+        except (KeyError, pebble.ConnectionError):
+            return False
+
+    def _render_config_file(self) -> str:
+        """Render the OAuth2 Proxy configuration file."""
+        oauth_integration = False
+
+        if self.oauth.is_client_created():
+            oauth_provider_info = self.oauth.get_provider_info()
+            oauth_integration = True
+
+        with open("templates/oauth2-proxy.toml.j2", "r") as file:
+            template = Template(file.read())
+
+        rendered = template.render(
+            authenticated_emails_file=ACCESS_LIST_EMAILS_PATH if self.config["authenticated-emails-list"] else None,
+            client_id=oauth_provider_info.client_id if oauth_integration else "default",
+            client_secret=oauth_provider_info.client_secret if oauth_integration else "default",
+            oauth_integration=oauth_integration,
+            oidc_issuer_url=oauth_provider_info.issuer_url if oauth_integration else None,
+            scopes=OAUTH_SCOPES,
         )
+        return rendered
+
+    def _restart_service(self) -> None:
+        """Build a new pebble layer and restart the service."""
+        self._container.add_layer(WORKLOAD_CONTAINER, self._oauth2_proxy_pebble_layer, combine=True)
+
+        try:
+            self._container.restart(WORKLOAD_CONTAINER)
+        except ChangeError as err:
+            logger.error(str(err))
+            self.unit.status = BlockedStatus(
+                "Failed to restart the container, please consult the logs"
+            )
+            return
 
     @log_event_handler(logger)
-    def _on_pebble_ready(self, event):
+    def _on_pebble_ready(self, event: PebbleReadyEvent) -> None:
         """Handle pebble ready event.
 
+            - open application port
+            - handle status and update config
+            - set workload version.
+
         Args:
-            event: The event triggered when the relation changed.
+            event: The event triggered when Pebble is ready for a workload.
         """
-        self._update(event)
+        self.model.unit.open_port(protocol="tcp", port=OAUTH2_PROXY_API_PORT)
+        self._handle_status_update_config(event)
+        self.unit.set_workload_version(OAUTH2_PROXY_VERSION)
```

Comment:
We can't be sure that this is the resource's version. I would suggest to get it by calling the oauth2-proxy CLI. No need to do it here, but let's add a TODO/issue/or just to your backlog.

Reply:
Good point, I'll switch it to call the cli